### PR TITLE
[1LP][RFR] Middleware conversion to widgetastic

### PR DESCRIPTION
**Pull request by hhovsepy**
{{pytest: cfme/tests//cloud/test_providers.py cfme/tests/infrastructure/test_providers.py -v}}

### CODE REVIEW

**Comment by hhovsepy** [comment]
random failures, timeouts, not test related

**Comment by hhovsepy** [comment]
Have included Middleware Server related changes PR https://github.com/ManageIQ/integration_tests/pull/5030 here.

**Comment by hhovsepy** [comment]
Could not remain Middleware Server untouched as it was related to my changes.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by quarckster** [comment]
@hhovsepy I would recommend to split this PR into several parts. I did the same in `cfme.intelligence.reports` widgetastic conversion: #5078, #5079, #5080, #5090, #5092, #5093, #5094

**Comment by hhovsepy** [comment]
Thanks @quarckster for hints.
@psav what is the general rules for dividing PRs for widgetastic conversation? I see that my PR is quite big and is all-in-one, but for middleware provider case, all objects are tightly coupled to each other, and dividing changes into several PRs will cause more mess.

**Comment by hhovsepy** [comment]
@jkrocil I see this common error in code:
"cfme/middleware/server.py:349: in step
    row = self.prerequisite_view.paginator.find_row_on_pages(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = PaginationPane()
obj = <cfme.middleware.provider.middleware_views.ServerAllView object at 0x7fac5ce1d850>
type = <class 'cfme.middleware.provider.middleware_views.ServerAllView'>

    def __get__(self, obj, type=None):
        if obj is None:  # class access
            return self
    
        # Cache on WidgetDescriptor
        if self not in obj._widget_cache:
            kwargs = copy(self.kwargs)
            try:
                kwargs['logger'] = create_child_logger(obj.logger, obj._desc_name_mapping[self])
            except AttributeError:
                pass
    
            args, kwargs = process_parameters(obj, self.args, kwargs)
>           if issubclass(self.klass, ParametrizedView):
E           TypeError: issubclass() arg 1 must be a class"

**Comment by izapolsk** [comment]
@hhovsepy, 
PaginationPane is function now.
https://github.com/ManageIQ/integration_tests/blob/master/widgetastic_manageiq/__init__.py#L1426
So, you just need to call it there instead of using View.nested.

Besides that, if you use BaseEntitiesView or ProviderView, it already includes paginator. So, you seems don't need it.

**Comment by hhovsepy** [comment]
There is such a failure which is not possible to reproduce while running from local machine:
"cfme/tests/middleware/test_middleware_domain_server.py:27: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cfme/middleware/server.py:307: in resume_server
    view = self.load_details(refresh=True)
cfme/middleware/server.py:196: in load_details
    view.browser.selenium.refresh()
/usr/lib/python2.7/site-packages/selenium/webdriver/remote/webdriver.py:614: in refresh
    self.execute(Command.REFRESH)
/usr/lib/python2.7/site-packages/selenium/webdriver/remote/webdriver.py:234: in execute
    response = self.command_executor.execute(driver_command, params)
/usr/lib/python2.7/site-packages/selenium/webdriver/remote/remote_connection.py:401: in execute
    return self._request(command_info[0], url, body=data)
/usr/lib/python2.7/site-packages/selenium/webdriver/remote/remote_connection.py:471: in _request
    resp = opener.open(request, timeout=self._timeout)
/usr/lib64/python2.7/urllib2.py:431: in open
    response = self._open(req, data)
/usr/lib64/python2.7/urllib2.py:449: in _open
    '_open', req)
/usr/lib64/python2.7/urllib2.py:409: in _call_chain
    result = func(*args)
/usr/lib64/python2.7/urllib2.py:1229: in http_open
    return self.do_open(httplib.HTTPConnection, req)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <urllib2.HTTPHandler instance at 0x7f1b48209878>
http_class = <class httplib.HTTPConnection at 0x7f1b6a00c4c8>
req = <selenium.webdriver.remote.remote_connection.Request instance at 0x7f1b480a8290>
http_conn_args = {}, host = '10.8.58.14:4911'
h = <httplib.HTTPConnection instance at 0x7f1b482096c8>
err = error(111, 'Connection refused')

    def do_open(self, http_class, req, **http_conn_args):
        """Return an addinfourl object for the request, using http_class.
    
            http_class must implement the HTTPConnection API from httplib.
            The addinfourl return value is a file-like object.  It also
            has methods and attributes including:
                - info(): return a mimetools.Message object for the headers
                - geturl(): return the original request URL
                - code: HTTP status code
            """
        host = req.get_host()
        if not host:
            raise URLError('no host given')
    
        # will parse host:port
        h = http_class(host, timeout=req.timeout, **http_conn_args)
        h.set_debuglevel(self._debuglevel)
    
        headers = dict(req.unredirected_hdrs)
        headers.update(dict((k, v) for k, v in req.headers.items()
                            if k not in headers))
    
        # We want to make an HTTP/1.1 request, but the addinfourl
        # class isn't prepared to deal with a persistent connection.
        # It will try to read all remaining data from the socket,
        # which will block while the server waits for the next request.
        # So make sure the connection gets closed after the (only)
        # request.
        headers["Connection"] = "close"
        headers = dict(
            (name.title(), val) for name, val in headers.items())
    
        if req._tunnel_host:
            tunnel_headers = {}
            proxy_auth_hdr = "Proxy-Authorization"
            if proxy_auth_hdr in headers:
                tunnel_headers[proxy_auth_hdr] = headers[proxy_auth_hdr]
                # Proxy-Authorization should not be sent to origin
                # server.
                del headers[proxy_auth_hdr]
            h.set_tunnel(req._tunnel_host, headers=tunnel_headers)
    
        try:
            h.request(req.get_method(), req.get_selector(), req.data, headers)
        except socket.error, err: # XXX what error?
            h.close()
>           raise URLError(err)
E           URLError: <urlopen error [Errno 111] Connection refused>"

**Comment by hhovsepy** [comment]
Another failure not possible to reproduce locally:
"cfme/middleware/server.py:308: in resume_server
    view.toolbar.power.item_select('Resume Server', handle_alert=True)
/usr/lib/python2.7/site-packages/widgetastic_patternfly/__init__.py:1383: in item_select
    self.open()
/usr/lib/python2.7/site-packages/widgetastic_patternfly/__init__.py:1319: in open
    self._verify_enabled()
/usr/lib/python2.7/site-packages/widgetastic_patternfly/__init__.py:1311: in _verify_enabled
    if not self.is_enabled:
/usr/lib/python2.7/site-packages/widgetastic_patternfly/__init__.py:1307: in is_enabled
    button = self.browser.element(self.BUTTON_LOCATOR, parent=self)
/usr/lib/python2.7/site-packages/widgetastic/browser.py:287: in element
    elements = self.elements(locator, *args, **kwargs)
/usr/lib/python2.7/site-packages/widgetastic/browser.py:738: in elements
    force_check_safe=force_check_safe)
/usr/lib/python2.7/site-packages/widgetastic/browser.py:257: in elements
    root_element = self.element(parent, parent=parent.locatable_parent)
/usr/lib/python2.7/site-packages/widgetastic/browser.py:287: in element
    elements = self.elements(locator, *args, **kwargs)
/usr/lib/python2.7/site-packages/widgetastic/browser.py:246: in elements
    locator = self._process_locator(locator)
/usr/lib/python2.7/site-packages/widgetastic/browser.py:199: in _process_locator
    return locator.__element__()
/usr/lib/python2.7/site-packages/widgetastic/widget.py:291: in __element__
    return self.parent_browser.element(locator)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <utils.appliance.implementations.ui.MiqBrowser object at 0x7f1b50945090>
locator = Locator(by='xpath', locator=u'.//div[contains(@class, "dropdown") and ./button[normalize-space(.)="Power" or normalize-space(@title)="Power"]]')
args = (), kwargs = {}, vcheck = None, elements = []

    def element(self, locator, *args, **kwargs):
        """Returns one :py:class:`selenium.webdriver.remote.webelement.WebElement`
    
            See: :py:meth:`elements`
    
            Returns:
                :py:class:`selenium.webdriver.remote.webelement.WebElement`
    
            Raises:
                :py:class:`selenium.common.exceptions.NoSuchElementException`
            """
        try:
            vcheck = self._locator_force_visibility_check(locator)
            if vcheck is not None:
                kwargs['check_visibility'] = vcheck
            elements = self.elements(locator, *args, **kwargs)
            if len(elements) > 1:
                visible_elements = [e for e in elements if self.is_displayed(e)]
                if visible_elements:
                    return visible_elements[0]
                else:
                    return elements[0]
            else:
                return elements[0]
        except IndexError:
>           raise NoSuchElementException('Could not find an element {}'.format(repr(locator)))
E           NoSuchElementException: Message: Could not find an element Locator(by='xpath', locator=u'.//div[contains(@class, "dropdown") and ./button[normalize-space(.)="Power" or normalize-space(@title)="Power"]]')"

**Comment by hhovsepy** [comment]
@jkrocil @izapolsk can you please take a look into this errors?

**Comment by izapolsk** [comment]
@hhovsepy, I think this happens because of
`view.browser.selenium.refresh()`
in load_details.
You call browser refresh and  return view. I suppose view isn't ready yet when you try to use it.
So, you have to wait until browser.refresh is over in load_details in this case.
I think this issue can be sorted out by ensure_page_safe call.

btw, many such views have refresh/reload button. You could use it instead if it is present.

**Comment by hhovsepy** [comment]
@izapolsk or the solution could be:
"view.browser.selenium.refresh()
 view.flush_widget_cache()"

**Comment by hhovsepy** [comment]
But the core problem is issue "https://github.com/ManageIQ/manageiq/issues/10333" which breaks MiQ instance and UI.

**Comment by izapolsk** [comment]
@hhovsepy, that's remarkable. however, please try ensure_page_safe if your issue won't disapear when miq issue gets fixed

**Comment by hhovsepy** [comment]
@izapolsk miq issue will not get fixed soon (it's 1 year old), can this PR be fully reviewed please?

**Comment by izapolsk** [comment]
sure, I'll try to review it today. It is quite big though.

File: cfme/middleware/deployment.py [hunk 1]
```diff
@@ -213,29 +218,46 @@ def deployment_in_rest(self):
...
+        self.view.entities.paginator.uncheck_all()
 
 
 @navigator.register(MiddlewareDeployment, 'Details')
 class Details(CFMENavigateStep):
+    VIEW = DeploymentDetailsView
     prerequisite = NavigateToSibling('All')
 
+    def am_i_here(self):
+        return match_page(summary='{} (Summary)'.format(self.obj.name))
```
Comment (izapolsk) [inline cfme/middleware/deployment.py]: ^^

File: cfme/middleware/deployment.py [hunk 2]
```diff
@@ -213,29 +218,46 @@ def deployment_in_rest(self):
...
+
     def step(self):
-        from cfme.web_ui import paginator
-        if paginator.page_controls_exist():
-            paginator.results_per_page(1000)
-        list_tbl.click_row_by_cells({'Deployment Name': self.obj.name,
-                                     'Server': self.obj.server.name})
+        self.prerequisite_view.toolbar.view_selector.select('List View')
```
Comment (izapolsk) [inline cfme/middleware/deployment.py]: ^^

**Comment by hhovsepy** [comment]
Thank you @izapolsk for review, looking ...

File: cfme/middleware/datasource.py [hunk 3]
```diff
@@ -104,19 +109,15 @@ def __init__(self, name, server, provider=None, appliance=None, **kwargs):
...
-                    _server = MiddlewareServer(provider=provider,
-                                               name=row.server.text,
-                                               hostname=row.host_name.text)
-                    datasources.append(MiddlewareDatasource(
-                        provider=provider,
-                        server=_server,
-                        name=row.datasource_name.text,
-                        hostname=row.host_name.text))
+        view = _get_datasources_page(provider=provider, server=server)
+        for _ in view.entities.paginator.pages():
```
[the remark this answers is not in the document]
Comment (hhovsepy) [reply cfme/middleware/datasource.py]: view.entities.get_all(): fails when there is not "Name" column.
    for row in view.entities.get_all():
File "/home/hhovsepy/git/cfme_tests3/widgetastic_manageiq/__init__.py", line 2348
in get_all
    return [self.parent.entity_class(parent=self, name=name) for name in self.entity_names]
File "/home/hhovsepy/git/cfme_tests3/widgetastic_manageiq/__init__.py", line 2437
in entity_names
    return [row.name.text for row in self.elements.rows()]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = TableRow(Table('//div[@id="gtl_div"]//table', column_widgets={}, assoc_column=None, rows_ignore_top=None, rows_ignore_bottom=None), 0)
attr = 'name'

    def __getattr__(self, attr):
        try:
            return self[self.table.ensure_normal(attr)]
        except KeyError:
>           raise AttributeError('Cannot find column {} in the table'.format(attr))
E           AttributeError: Cannot find column name in the table

File: cfme/middleware/datasource.py [hunk 4]
```diff
@@ -245,41 +241,92 @@ def datasource_in_rest(self):
...
-        list_tbl.click_row_by_cells({
-            'Datasource Name': self.obj.name,
-            'Server': self.obj.server.name,
-            'Host Name': self.obj.hostname
-        })
+    def am_i_here(self):
+        return match_page(summary='{} (Summary)'.format(self.obj.name))
+
+    def step(self, *args, **kwargs):
+        self.prerequisite_view.toolbar.view_selector.select('List View')
```
[the remark this answers is not in the document]
Comment (hhovsepy) [reply cfme/middleware/datasource.py]: for datasources list, there is not column "Name", it is "Datasource Name", and also I need to search here by "Server Name",
get_entity/get_all fails here

[the remark this answers is not in the document]
Comment (izapolsk) [reply cfme/middleware/datasource.py]: you can easily override that in ListEntity class. 
Besides that,  it should work well in Grid and Tile mode.

File: cfme/middleware/datasource.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (izapolsk) [reply cfme/middleware/datasource.py]: you can easily override that in ListEntity class.
Besides that, it should work well in Grid and Tile mode.
Please let me know if you need any examples

[the remark this answers is not in the document]
Comment (hhovsepy) [reply cfme/middleware/datasource.py]: yes please

[the remark this answers is not in the document]
Comment (izapolsk) [reply cfme/middleware/datasource.py]: I guess it should look like 
```
DatasourceEntitiesView(BaseEntitiesView):
    """Entities on the main list page"""
    @entities.register('List View')
    class ListView(EntitiesConditionalView):
        elements = Table(locator='//div[@id="gtl_div"]//table')

        @property
        def entity_names(self):
            """ looks for entities and extracts their names

            Returns: all current page entities
            """
            try:
                return [row.database_name.text for row in self.elements.rows()]
            except NoSuchElementException:
                return []
    
    title = Text(TITLE_LOCATOR)
    table = Table(LIST_TABLE_LOCATOR)
    search = View.nested(Search)
    # element attributes changed from id to class in upstream-fine+, capture both with locator
    flash = FlashMessages(FLASH_MESSAGE_LOCATOR)
```

**Comment by hhovsepy** [comment]
@izapolsk have changed according to comments, besides the entities searching as we agreed in chat

File: cfme/middleware/datasource.py [hunk 5]
```diff
@@ -245,41 +237,87 @@ def datasource_in_rest(self):
...
+
+    def wait_for_exists(self):
+        """Wait for the datasource to be refreshed"""
+        view = navigate_to(self, 'All')
+
+        def refresh():
+            if self.provider:
+                self.provider.refresh_provider_relationships()
+            view.browser.selenium.refresh()
+            view.flush_widget_cache()
```
[the remark this answers is not in the document]
Comment (hhovsepy) [reply cfme/middleware/datasource.py]: I think this is the workaround of problem when page is refreshed and view is not loaded yet, isn't it?

File: cfme/middleware/messaging.py [hunk 6]
```diff
@@ -220,27 +222,45 @@ def messaging_in_rest(self):
...
+        self.view.entities.paginator.check_all()
+        self.view.entities.paginator.uncheck_all()
 
 
 @navigator.register(MiddlewareMessaging, 'Details')
 class Details(CFMENavigateStep):
+    VIEW = MessagingDetailsView
     prerequisite = NavigateToSibling('All')
 
+    def am_i_here(self):
```
Comment (izapolsk) [inline cfme/middleware/messaging.py]: ^^

File: cfme/middleware/provider/__init__.py [hunk 7]
```diff
@@ -181,55 +143,60 @@ class Timelines(CFMENavigateStep):
...
     prerequisite = NavigateToSibling('Details')
 
     def step(self):
-        sel.click(InfoBlock.element('Relationships', 'Middleware Messagings'))
+        self.prerequisite_view.contents.relationships.click_at('Middleware Messagings')
 
 
 @navigator.register(MiddlewareProvider, 'TopologyFromDetails')
 class TopologyFromDetails(CFMENavigateStep):
     prerequisite = NavigateToSibling('Details')
```
Comment (izapolsk) [inline cfme/middleware/provider/__init__.py]: I don't see view statement here

File: cfme/middleware/provider/__init__.py [hunk 8]
```diff
@@ -466,34 +379,31 @@ def undeploy(self):
         """
         Clicks on "Undeploy" menu item and verifies message shown
         """
-        self.load_details()
-        operations_btn("Undeploy", invokes_alert=True)
-        sel.handle_alert()
-        flash.assert_success_message('Undeployment initiated for selected deployment(s)')
+        view = self.load_details()
+        view.toolbar.operations.item_select("Undeploy", handle_alert=True)
+        view.flash \
```
Comment (izapolsk) [inline cfme/middleware/provider/__init__.py]: please get rid of blackslash here

File: cfme/middleware/provider/__init__.py [hunk 9]
```diff
@@ -466,34 +379,31 @@ def undeploy(self):
...
 
     def restart(self):
         """
         Clicks on "Restart" menu item and verifies message shown
         """
-        self.load_details()
-        operations_btn("Restart", invokes_alert=True)
-        sel.handle_alert()
-        flash.assert_success_message('Restart initiated for selected deployment(s)')
+        view = self.load_details()
```
Comment (izapolsk) [inline cfme/middleware/provider/__init__.py]: honestly I don't think load_details() method is necessary at all. it does almost the same as navigate_to(obj, 'Details')

File: cfme/middleware/deployment.py [hunk 10]
```diff
@@ -213,29 +215,45 @@ def deployment_in_rest(self):
...
+        self.view.entities.paginator.check_all()
+        self.view.entities.paginator.uncheck_all()
 
 
 @navigator.register(MiddlewareDeployment, 'Details')
 class Details(CFMENavigateStep):
+    VIEW = DeploymentDetailsView
     prerequisite = NavigateToSibling('All')
 
+    def am_i_here(self):
```
[the remark this answers is not in the document]
Comment (hhovsepy) [reply cfme/middleware/deployment.py]: eclipse search problem :)

[the remark this answers is not in the document]
Comment (izapolsk) [reply cfme/middleware/deployment.py]: *it happens :)

File: cfme/middleware/provider/__init__.py (hunk 7 above)
Comment (hhovsepy) [reply cfme/middleware/provider/__init__.py]: Topology is not Middleware Specific, can we leave it for later?

Comment (izapolsk) [reply cfme/middleware/provider/__init__.py]: sure, please add appropriate comment there

File: cfme/middleware/provider/__init__.py (hunk 9 above)
Comment (hhovsepy) [reply cfme/middleware/provider/__init__.py]: Here 'Deployable' is a superclass, and load_details() method is overridden in subclasses with necessary calls to "navigate_to()" and other specific calls.
In general, I keep a style to call "navigate_to()" only from "load_details()" method of particular class, for the rest self.load_details() is called.

**Comment by hhovsepy** [comment]
@izapolsk so far have fixed all comments,
this PR becomes blocker for me in adding new tests

**Comment by izapolsk** [comment]
@hhovsepy, I see, I'll try to review it asap

File: cfme/tests/middleware/test_middleware_provider.py [hunk 11]
```diff
@@ -25,56 +25,62 @@
...
+    view = provider.create_view(MiddlewareProviderAddView)
+    assert not view.add.active
 
 
 def test_add_cancelled_validation():
     """Tests that the flash message is correct when add is cancelled."""
     prov = HawkularProvider()
     prov.create(cancel=True)
-    flash.assert_message_match('Add of Middleware Provider was cancelled by the user')
+    view = prov.browser.create_view(MiddlewareProviderAddView)
```
Comment (izapolsk) [inline cfme/tests/middleware/test_middleware_provider.py]: maybe ProviderAllView ?

File: cfme/tests/middleware/test_middleware_provider.py [hunk 12]
```diff
@@ -25,56 +25,62 @@
...
     )
-    navigate_to(prov, 'Add')
-    fill(prov.properties_form, {"type_select": "Hawkular", "hostname_text": "test",
-                                "name_text": "test", "port_text": "8080"})
-    fill(cred.form, cred)
-    sel.wait_for_ajax()
-    soft_assert(not form_buttons.validate.can_be_clicked)
-    soft_assert(not prov.add_provider_button.can_be_clicked)
-    soft_assert(cred.form.default_verify_secret.angular_help_block == "Passwords do not match")
+    add_view = navigate_to(provider, 'Add')
```
Comment (izapolsk) [inline cfme/tests/middleware/test_middleware_provider.py]: you already have prepared provider object. You can copy this object, override default endpoint credentials and call its create() method. You won't need to play with all these fills then

File: cfme/tests/middleware/test_middleware_provider.py [hunk 13]
```diff
@@ -25,56 +25,62 @@
...
 
 @pytest.mark.meta(blockers=['GH#ManageIQ/manageiq:8241'])
 @pytest.mark.usefixtures('has_no_middleware_providers')
 def test_provider_add_with_invalid_port(provider, soft_assert):
     """ Tests provider add with bad port (non-numeric)"""
-    provider.port = fauxfactory.gen_alpha(6)
-    with error.expected(FlashMessageException):
-            provider.create(validate_credentials=False)
-    soft_assert(not provider.add_provider_button.can_be_clicked)
-    soft_assert(provider.properties_form.port_text.angular_help_block ==
```
Comment (izapolsk) [inline cfme/tests/middleware/test_middleware_provider.py]: ^^

**Comment by hhovsepy** [comment]
@izapolsk thanks for catch, have fixed all

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by psav** [comment]
Tested this locally to be sure, and seems fine on other providers - hold onto your hats people

**Pull request merged**